File: app/src/main/java/com/rexo/marketplace/utils/ErrorUtils.kt
package com.rexo.marketplace.utils

/**
 * Utility for sanitizing error messages before displaying them to users.
 * Strips technical details like URLs, API keys, JSON errors, and Supabase internals.
 * Returns clean, user-friendly messages.
 */
object ErrorUtils {

    private val URL_PATTERN = Regex("https?://[^\\s\"']+", RegexOption.IGNORE_CASE)
    private val API_KEY_PATTERN = Regex("(apikey|key|token|secret|authorization)[=:\\s]*[^\\s\"',}]+", RegexOption.IGNORE_CASE)
    private val JSON_DETAIL_PATTERN = Regex("\\{[^}]*\\}")

    /**
     * Sanitize a raw error message for user display.
     * - Strips URLs (http/https patterns)
     * - Strips API keys and tokens
     * - Strips JSON details
     * - Returns user-friendly messages based on recognized keywords
     * - Default: "Something went wrong. Please try again."
     */
    fun sanitizeErrorMessage(rawMessage: String?): String {
        if (rawMessage.isNullOrBlank()) {
            return "Something went wrong. Please try again."
        }

        val lower = rawMessage.lowercase()

        // Network / connectivity issues
        if (lower.contains("timeout") || lower.contains("timed out")) {
            return "Connection timed out. Please check your internet and try again."
        }
        if (lower.contains("unable to resolve host") || lower.contains("no address associated") ||
            lower.contains("network") || lower.contains("connect")) {
            return "Network error. Please check your internet connection."
        }

        // Authentication issues
        if (lower.contains("not authenticated") || lower.contains("unauthorized") ||
            lower.contains("sign in") || lower.contains("login")) {
            return "Please sign in to continue."
        }
        if (lower.contains("invalid login") || lower.contains("invalid credentials")) {
            return "Invalid email or password. Please try again."
        }

        // Table/schema issues (should not happen in production, but handle gracefully)
        if (lower.contains("could not find") && lower.contains("table")) {
            return "Service temporarily unavailable. Please try again later."
        }
        if (lower.contains("schema")) {
            return "Service temporarily unavailable. Please try again later."
        }

        // JSON / deserialization errors
        if (lower.contains("json") || lower.contains("serializ") || lower.contains("deserializ") ||
            lower.contains("unexpected") || lower.contains("parsing")) {
            return "Something went wrong. Please try again."
        }

        // Permission issues
        if (lower.contains("permission") || lower.contains("forbidden") || lower.contains("denied")) {
            return "You don't have permission to perform this action."
        }

        // Not found
        if (lower.contains("not found") || lower.contains("no rows")) {
            return "The requested item was not found."
        }

        // Rate limiting
        if (lower.contains("rate limit") || lower.contains("too many requests")) {
            return "Too many requests. Please wait a moment and try again."
        }

        // If the message contains URLs or API keys, strip and return generic
        if (URL_PATTERN.containsMatchIn(rawMessage) || API_KEY_PATTERN.containsMatchIn(rawMessage)) {
            return "Something went wrong. Please try again."
        }

        // If message is short and clean (no technical jargon), allow it through
        val cleaned = rawMessage
            .let { URL_PATTERN.replace(it, "") }
            .let { API_KEY_PATTERN.replace(it, "") }
            .let { JSON_DETAIL_PATTERN.replace(it, "") }
            .trim()

        if (cleaned.length < 80 && !cleaned.contains("Exception") &&
            !cleaned.contains("supabase", ignoreCase = true) &&
            !cleaned.contains("postgrest", ignoreCase = true) &&
            !cleaned.contains("\$[")) {
            return cleaned.ifBlank { "Something went wrong. Please try again." }
        }

        return "Something went wrong. Please try again."
    }
}

File: lib/core/utils/error_utils.dart
/// Utility for sanitizing error messages before displaying to users.
///
/// NEVER show raw exception messages (PostgrestException, Supabase URLs,
/// API keys, stack traces) to end users. Always use this utility to
/// convert errors into clean, user-friendly messages.
class ErrorUtils {
  ErrorUtils._();

  /// Convert a raw error object into a user-friendly message.
  ///
  /// Strips Supabase URLs, API keys, technical details, and
  /// PostgreSQL error codes. Returns a clean message suitable
  /// for display in the UI.
  static String sanitize(Object? error) {
    if (error == null) return 'Something went wrong. Please try again.';

    final message = error.toString().toLowerCase();

    // TEMP DEBUG (v4): raw error for jobs (campaigns is_job query). Remove after fix.
    return 'DBG4: ${error.toString()}';

    // ignore: dead_code
    // Network/connection errors
    if (message.contains('socketexception') ||
        message.contains('connection refused') ||
        message.contains('network') ||
        message.contains('timeout') ||
        message.contains('no internet')) {
      return 'No internet connection. Please check your network and try again.';
    }

    // Authentication errors
    if (message.contains('not authenticated') ||
        message.contains('jwt expired') ||
        message.contains('invalid token') ||
        message.contains('session expired')) {
      return 'Your session has expired. Please sign in again.';
    }

    if (message.contains('invalid login') ||
        message.contains('invalid credentials') ||
        message.contains('email not confirmed')) {
      return 'Invalid email or password. Please try again.';
    }

    // Permission errors
    if (message.contains('permission denied') ||
        message.contains('not authorized') ||
        message.contains('rls') ||
        message.contains('row-level security') ||
        message.contains('policy')) {
      return 'You do not have permission to perform this action.';
    }

    // Recursion error (the critical bug)
    if (message.contains('infinite recursion')) {
      return 'A server configuration issue occurred. Please contact support.';
    }

    // Not found
    if (message.contains('not found') || message.contains('no rows')) {
      return 'The requested data was not found.';
    }

    // Handle/username uniqueness errors
    if (message.contains('handle') &&
        (message.contains('unique') ||
            message.contains('duplicate') ||
            message.contains('already exists'))) {
      return 'This handle is already taken. Please choose a different one.';
    }

    // Duplicate/conflict
    if (message.contains('duplicate') ||
        message.contains('already exists') ||
        message.contains('unique constraint')) {
      return 'This item already exists. Please try a different value.';
    }

    // Insufficient balance
    if (message.contains('insufficient balance') ||
        message.contains('insufficient funds')) {
      return 'Insufficient balance for this transaction.';
    }

    // Insufficient stock
    if (message.contains('insufficient stock')) {
      return 'This product is out of stock.';
    }

    // Generic PostgrestException pattern - strip technical details
    if (message.contains('postgrestexception') ||
        message.contains('postgresql') ||
        message.contains('supabase') ||
        message.contains('42p') ||
        message.contains('https://')) {
      return 'Something went wrong. Please try again.';
    }

    // If the message is short and doesn't contain URLs or technical info, allow it
    final raw = error.toString();
    if (raw.length < 100 &&
        !raw.contains('http') &&
        !raw.contains('Exception') &&
        !raw.contains('Error(') &&
        !raw.contains('stacktrace')) {
      return raw;
    }

    // Default fallback
    return 'Something went wrong. Please try again.';
  }
}
